File: 2021/solutions/day2.py
from pathlib import Path

def part_1():
  """Simple part 1"""
  x, depth = 0, 0
  for instruction, value in _iter_instructions():
    if instruction == "forward":
      x += value
    elif instruction == "down":
      depth += value
    elif instruction == "up":
      depth -= value
  print(f"PART 1: Location is: {x} x {depth} = {x * depth}")


def part_2():
  """Tricky part 2"""
  hor, depth, aim = 0, 0, 0
  for instruction, x in _iter_instructions():
    if instruction == "forward":
      hor += x
      depth += (aim * x)
    elif instruction == "down":
      aim += x
    elif instruction == "up":
      aim -= x
  print(f"PART 2: Location is: {hor} x {depth} = {hor * depth}")


def _iter_instructions():
  input_file = Path(__file__).parent.parent / "data" / "day2.txt"
  with input_file.open() as f:
    for line in f:
      instruction, value = line.strip().split()
      yield instruction, int(value)


if __name__ == "__main__":
  part_1()
  part_2()

File: 2020/solutions/day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

// Policy struct used to check passwords
type Policy struct {
	char  string
	lower int
	upper int
}

func (p *Policy) isValidByBounds(password string) bool {
	occurrences := strings.Count(password, p.char)
	return (occurrences >= p.lower) && (occurrences <= p.upper)
}

func (p *Policy) isValidByLocation(password string) bool {
	match1 := password[p.lower-1] == p.char[0]
	match2 := password[p.upper-1] == p.char[0]
	return (match1 || match2) && !(match1 && match2)
}

// Useful for passing out of loadData
type PolicyPasswordPair struct {
	policy   Policy
	password string
}

func main() {
	part1()
	part2()
}

func part1() {
	// Having a go of using channels to pass data around
	channel := make(chan PolicyPasswordPair)
	go loadData(channel)
	nValid := 0
	for pair := range channel {
		if pair.policy.isValidByBounds(pair.password) {
			nValid++
		}
	}
	fmt.Printf("PART 1: %d valid passwords\n", nValid)
}

func part2() {
	nValid := 0
	channel := make(chan PolicyPasswordPair)
	go loadData(channel)
	for pair := range channel {
		if pair.policy.isValidByLocation(pair.password) {
			nValid++
		}
	}
	fmt.Printf("PART 2: %d valid passwords\n", nValid)
}

func loadData(channel chan<- PolicyPasswordPair) {
	path, _ := filepath.Abs(filepath.Join("data", "day2.txt"))
	file, err := os.Open(path)
	check(err)
	scanner := bufio.NewScanner((file))
	re := regexp.MustCompile("[- ]")
	for scanner.Scan() {
		line := strings.Split(scanner.Text(), ": ")
		parts := re.Split(line[0], 3)
		policy := Policy{char: parts[2], lower: atoi(parts[0]), upper: atoi(parts[1])}
		pair := PolicyPasswordPair{policy: policy, password: line[1]}
		channel <- pair
	}
	close(channel)
}

// Utility to simply ignore errors on conversion
func atoi(str string) int {
	val, _ := strconv.Atoi(str)
	return val
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}
